indico/web/client/js/react/components/DateRangePicker.jsx

```diff
@@ -0,0 +1,253 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2024 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import moment from 'moment';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import DatePickerCalendar from 'indico/react/components/DatePickerCalendar';
+import {FinalField, validators as v} from 'indico/react/forms';
+import {Param, Translate} from 'indico/react/i18n';
+import {DateRange} from 'indico/utils/date';
+import {formatDate, ISO_FORMAT} from 'indico/utils/date_format';
+import {fromISOLocalDate} from 'indico/utils/date_parser';
+
+import 'indico/custom_elements/ind_date_picker';
+
+const INVALID = '__invalid__';
+
+export default function DateRangePicker({
+  value,
+  label,
+  format,
+  rangeStartLabel,
+  rangeEndLabel,
+  rangeStartMin,
+  rangeStartMax,
+  rangeEndMin,
+  rangeEndMax,
+  min,
+  max,
+  readOnly,
+  disabled,
+  startDisabled,
+  endDisabled,
+  onChange,
+}) {
+  format ??= moment.localeData().longDateFormat('L');
+
+  if (min) {
+    rangeStartMin = rangeEndMin = min;
+  }
+
+  if (max) {
+    rangeEndMax = rangeEndMax = max;
+  }
+
+  function handleChange(ev) {
+    const picker = ev.currentTarget;
+    const [start, end] = picker.dateRange;
+    const rangeStartInput = picker.querySelector('[data-range-start]');
+    const rangeEndInput = picker.querySelector('[data-range-end]');
+
+    const startInvalid = rangeStartInput.value && !start;
+    const endInvalid = rangeEndInput.value && !end;
+
+    onChange({
+      // ISO local dates
+      startDate: startInvalid ? INVALID : formatDate(ISO_FORMAT, start) || null,
+      endDate: endInvalid ? INVALID : formatDate(ISO_FORMAT, end) || null,
+    });
+  }
+
+  return (
+    <ind-date-range-picker
+      range-start={fromISOLocalDate(value.startDate)?.toDateString() || ''}
+      range-end={fromISOLocalDate(value.endDate)?.toDateString() || ''}
+      range-start-min={rangeStartMin}
+      range-start-max={rangeStartMax}
+      range-end-min={rangeEndMin}
+      range-end-max={rangeEndMax}
+      start-disabled={startDisabled || undefined}
+      end-disabled={endDisabled || undefined}
```

Comment:
Currently we only need `startDisabled`, but I added `endDisabled` for the sake of completeness. However, I noticed when testing it that it's actually more complicated since we'd also need to prevent jumping to end date selection mode in that case, and close the picker instead. It's probably faster if you add this logic since it probably involves touching the state machine for the transitions.

Otherwise I'd also be OK with only allowing to lock the start date and hope we don't end up needing a locked end date at some point in the future.

Reply:
I've made changes to the state machine and the React component so that you can now disable either end of the range. I've added a sanity check that the value for the disabled end of the range must be present for the flag to take effect. Let me know if this is what you had in mind.

Reply:
Alternatively, maybe I could make it emit a warning when the flag is passed but value is missing.

Reply:
haven't tested it yet but looks very nice and just what i had in mind

Reply:
Just tested it and works as expected :)

And I think a warning makes sense in that case, because IMHO code should not disable a side when there's no value for it